public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/PyroscopeAppIntegration.tsx

```diff
@@ -0,0 +1,133 @@
+const pyroscopeDatasourceSettingsByUid: Record<string, Promise<PyroscopeDatasourceSettings>> = {};
+
+export type Props = QueryEditorProps<PyroscopeDataSource, Query, PyroscopeDataSourceOptions>;
+
+/** A subset of the app settings that are relevant for this integration */
+type PyroscopeAppSettings = {
+  enabled: boolean;
+  jsonData: {
+    backendUrl: string;
+    basicAuthUser: string;
+  }
+}
+
+/** A subset of the datasource settings that are relevant for this integration */
+type PyroscopeDatasourceSettings = {
+  url: string;
+  basicAuthUser: string | number; // This might be configured as a number, so we will want to ensure compare for string-equivalence
+}
+
+export function PyroscopeAppIntegration(props: Props) {
+
+  const [appPlugin, setAppPlugin] = useState<PyroscopeAppSettings>()
+  const [datasource, setDatasource] = useState<PyroscopeDatasourceSettings>()
+
+  const {datasource:{uid: datasourceUid}, query, range} = props;
+
+  useEffect(()=>{
+
+    if (pyroscopeAppSettings == null) {
+      pyroscopeAppSettings = getBackendSrv().get<PyroscopeAppSettings>(`/api/plugins/${PYROSCOPE_APP_ID}/settings`)
+    }
+
+    pyroscopeAppSettings.then(setAppPlugin);
+    
+  }, [])
+
+  useEffect(()=>{
+    let datasourceSettings = pyroscopeDatasourceSettingsByUid[datasourceUid];
+
+    if (datasourceSettings == null) {
+      datasourceSettings = getBackendSrv().get<PyroscopeDatasourceSettings>(`/api/datasources/uid/${datasourceUid}`)
+      pyroscopeDatasourceSettingsByUid[datasourceUid] = datasourceSettings;
+    }
+
+    datasourceSettings.then(setDatasource);
+
+  }, [datasourceUid])
+
+  const queryParam = useMemo(()=>{
+
+    return generateQueryParams(query, range);
+  }, [query, range])
+
+  const profilesAppLink = useMemo(()=>isPyroscopeDatasourceCompatibleWithPlugin(datasource, appPlugin) &&
+    <LinkButton variant='secondary' icon='external-link-alt' tooltip={'Open query in Profiles App'} target='_blank' href={`/a/${PYROSCOPE_APP_ID}/single?${queryParam}`}>Profiles App</LinkButton>
+  , [appPlugin, datasource, queryParam])
+
+  return profilesAppLink || null;
+}
+
+export function isPyroscopeDatasourceCompatibleWithPlugin(datasource?: PyroscopeDatasourceSettings, appPlugin?: PyroscopeAppSettings) {
+  if (!appPlugin || !datasource) {
+    return false;
+  }
+
+  return (
+    datasource.url === appPlugin.jsonData.backendUrl
+    &&
+    String(datasource.basicAuthUser) === String(appPlugin.jsonData.basicAuthUser)
+  )
+}
+
+
+function stringifyRawTimeRangePart(rawTimeRangePart: DateTime | string) {
+  if (typeof rawTimeRangePart === 'string') {
+    return rawTimeRangePart;
+  }
+
+  // The `unix` result as a string is compatible with Pyroscope's range part format
+  return Math.round(rawTimeRangePart.unix()).toString();
+}
+
+export function translateGrafanaTimeRangeToPyroscope(timeRange: TimeRange) {
+  const from = stringifyRawTimeRangePart(timeRange.raw.from);
```

Comment:
The `raw` part can also contain strings like `-1h`. Is pyroscope going to understand that?

Reply:
Fortunately yes! 
However, I have only encountered things such as: `now-1h` -- do we also produce cases of `[timestamp] - 1h` or are you only referring to the _relative-to-now_ subtractions?

Reply:
I think it should be only relative to now, don't think I have ever seen `[timestamp] - 1h`.